pkg/apis/azure/validation/credentials.go

```diff
@@ -0,0 +1,156 @@
+		}
+
+		value, exists := secret.Data[dataKey]
+		if !exists {
+			allErrs = append(allErrs, field.Required(fldPath.Key(dataKey),
+				fmt.Sprintf("missing required field %q in secret %s", dataKey, secretKey)))
+			continue
+		}
+		if len(value) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Key(dataKey), "",
+				fmt.Sprintf("field %q cannot be empty in secret %s", dataKey, secretKey)))
+		}
+	}
+
+	return allErrs
+}
+
+// validateFormats validates the format of credential values
+func (cm *CredentialMapping) validateFormats(secret *corev1.Secret, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	secretKey := fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)
+
+	for dataKey, spec := range cm.Fields {
+		value, exists := secret.Data[dataKey]
+		if !exists || len(value) == 0 {
+			continue
+		}
+
+		// Validate GUID format
+		if spec.IsGUID && !guidRegex.Match(value) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Key(dataKey), "(hidden)",
+				fmt.Sprintf("field %q must be a valid GUID in secret %s", dataKey, secretKey)))
+		}
+
+		// Validate no leading/trailing whitespace
+		valueStr := string(value)
+		if strings.TrimSpace(valueStr) != valueStr {
+			allErrs = append(allErrs, field.Invalid(fldPath.Key(dataKey), "(hidden)",
+				fmt.Sprintf("field %q must not contain leading or trailing whitespace in secret %s", dataKey, secretKey)))
+		}
+	}
+
+	return allErrs
+}
+
+// validateNoUnexpected validates that no unexpected fields are present in the secret
+func (cm *CredentialMapping) validateNoUnexpected(secret *corev1.Secret, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	secretKey := fmt.Sprintf("%s/%s", secret.Namespace, secret.Name)
+
+	allowedKeys := sets.New[string]()
+	for dataKey := range cm.Fields {
+		allowedKeys.Insert(dataKey)
+	}
+
+	for key := range secret.Data {
+		if !allowedKeys.Has(key) {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Key(key),
+				fmt.Sprintf("unexpected field %q in secret %s", key, secretKey)))
+		}
+	}
+
+	return allErrs
+}
+
+// validateImmutable validates that immutable fields haven't changed
+func (cm *CredentialMapping) validateImmutable(newSecret, oldSecret *corev1.Secret, resourceType string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	secretKey := fmt.Sprintf("%s/%s", newSecret.Namespace, newSecret.Name)
+
+	for dataKey, spec := range cm.Fields {
+		if !spec.IsImmutable {
+			continue
+		}
+
+		if !bytes.Equal(newSecret.Data[dataKey], oldSecret.Data[dataKey]) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Key(dataKey), "(hidden)",
+				fmt.Sprintf("field %q must not be changed for existing %s in secret %s", dataKey, resourceType, secretKey)))
+		}
+	}
+
+	return allErrs
+}
+
+// ValidatePredefinedValues validates that a field contains one of the predefined values
+func ValidatePredefinedValues(secret *corev1.Secret, dataKey string, allowedValues []string, fldPath *field.Path) field.ErrorList {
```

Comment:
I have the feeling that this could also probably be part of the FieldSpec and the function part of the `Validate` function. What was the reason for leaving it out ?

Reply:
My original intend was to separate it since it's only used for the DNS secrets. But it's indeed cleaner to integrate it into the overall `Validate` function.